rust/auction/program/src/processor/cancel_bid.rs

```diff
@@ -133,6 +139,28 @@ pub fn cancel_bid(
     // and update cancelled bids if auction still active
     let mut auction_extended = AuctionDataExtended::from_account_info(accounts.auction_extended)?;
 
+    // Obtain correct BidState instance
+    let (mut bid_state, mut bid_state_data_acc) =
+        if let Some(bid_state_data) = accounts.bid_state_data {
+            if *bid_state_data.key
+                != auction_extended
+                    .bid_state_data
+                    .ok_or(ProgramError::InvalidArgument)?
+            {
+                return Err(ProgramError::InvalidArgument);
+            }
+
+            let bid_state_data_acc: BidStateData =
+                try_from_slice_unchecked(&bid_state_data.data.borrow_mut())?;
+
+            (
+                bid_state_data_acc.state.clone(),
+                Some(bid_state_data_acc.clone()),
+            )
+        } else {
+            (auction.bid_state.clone(), None)
```

Comment:
Are you sure that we have to `clone()` bid state here ?
I mean we have to mutate data in `auction` struct